coral-schema/src/main/java/com/linkedin/coral/schema/avro/RelToAvroSchemaConverter.java

```diff
@@ -442,6 +436,22 @@ public RexNode visitCall(RexCall rexCall) {
        * For SqlUserDefinedFunction and SqlOperator RexCall, no need to handle it recursively
        * and only return type of udf or sql operator is relevant
        */
+
+      /**
+       * If the return type of RexCall is based on the ordinal of its input argument
+       * and the corresponding input argument refers to a field from the input schema,
+       * use the field's schema as is.
+       */
+      if (rexCall.getOperator().getReturnTypeInference() instanceof OrdinalReturnTypeInferenceV2) {
+        int index = ((OrdinalReturnTypeInferenceV2) rexCall.getOperator().getReturnTypeInference()).getOrdinal();
+        RexNode operand = rexCall.operands.get(index);
+
+        if (operand instanceof RexInputRef) {
+          appendRexInputRefField((RexInputRef) operand);
+          return rexCall;
+        }
+      }
+
       RelDataType fieldType = rexCall.getType();
       boolean isNullable = SchemaUtilities.isFieldNullable(rexCall, inputSchema);
```

Comment:
Is this the reason `OrdinalReturnTypeInference` does not work out of the box?

Reply:
`[OrdinalReturnTypeInference](https://github.com/linkedin/linkedin-calcite/blob/li-1.21.0/core/src/main/java/org/apache/calcite/sql/type/OrdinalReturnTypeInference.java#L25)` cannot be directly used because the private field 'ordinal' has no public accessor method. This class supports type derivation via method `RelDataType inferReturnType(SqlOperatorBinding opBinding)`. To leverage this API, we need to make coral-schema complaint with Coral IR and enable type derivation in coral-schema, similar to our ongoing work in coral-spark & coral-trino

Reply:
So once we append a field, what gets used at the end to infer its type if not `inferReturnType`?

Reply:
we append the field AFTER extracting the ordinal [here](https://github.com/linkedin/coral/pull/481/files#diff-a2e05f91f268adfd9edbe9dda395197fad075217aaac529dbee5d794a8f6866cR446) and deriving its field type [here](https://github.com/linkedin/coral/pull/481/files#diff-a2e05f91f268adfd9edbe9dda395197fad075217aaac529dbee5d794a8f6866cR559) 

Since this is a rexShuttle, the field types from the inputSchema are available in the RelNode representation. 

`SchemaUtilities.isFieldNullable(rexCall, inputSchema)` derives incorrect nullability (always nullable) for the top level fields and `SchemaUtilities.makeNullable` [here](https://github.com/linkedin/coral/pull/481/files#diff-af50d4a6ecbf662497acd9da8f31d6fb50300e833ddad2716ec6814a9732d243R89) derives incorrect nullability (always nullable) for the inner fields. 

By introducing the `if` condition above in line 445, we avoid those codepaths